timemachine/cpp/src/kernels/k_exchange.cu

```diff
@@ -747,16 +747,19 @@ template void __global__ k_setup_destination_weights_for_targeted<double>(
 );
 
 void __global__ k_adjust_sample_idxs(
+    const int total_proposals,
     const int batch_size,
+    const int *__restrict__ noise_offset,           // [1]
     const int *__restrict__ targeting_inner_volume, // [batch_size]
     const int *__restrict__ inner_count,            // [1]
-    const int *__restrict__ partitioned_indices,    // [inner_count]
+    const int *__restrict__ partitioned_indices,    // [num_samples]
     int *__restrict__ sample_idxs                   // [batch_size]
 ) {
-    const int local_inner_count = inner_count[0];
+    const int current_offset = *noise_offset;
+    const int local_inner_count = *inner_count;
     int idx = blockIdx.x * blockDim.x + threadIdx.x;
 
-    while (idx < batch_size) {
+    while (idx < batch_size && idx + current_offset < total_proposals) {
```

Comment:
trying to follow the out of bounds logic here:

`partitioned_indices is of size num_samples`
`sample_idxs` is of size `batch_size`

So while `sample_idxs[idx]` will never be out of bounds, the accessed element `before` may be out of bounds when used to index into `before+offset`?

Reply:
The accessed element in `partitioned_indices[before + offset]` was what could be wrong. If `before == num_outer + 1`then it would crash. The value of `before` could be larger than expected since upstream we don't recompute the `local_inner_count` value once we have gone beyond the total number of proposals.

Reply:
is checking `idx + current_offset < total_proposals` equivalent to checking `before + offset < num_samples`?

Reply:
(the latter is more obvious that it prevents an out of bounds read) - less clear still on how it's guaranteed / implied that `idx + current_offset < num_samples`

Reply:
No, `num_samples` is `batch_size`. `total_proposals` is equal to `proposals_per_move` in the TIBD mover and the current offset is the number proposals already made.